### NumPy
### numpy is python's package for doing math that is more advanced than +-*/

### This includes special functions like cosine, exponential, sqrt, ...

### On top of this we can use numpy to generate samples from many types of random variables

### numpy also has a powerful data type to define vectors, matrices, and tensors

### With these data types numpy also allows us to do linear algebra - matrix multiplication and matrix-vector solutions
### a matrix is 2d vector ,tensor is any array thats a higher dimension than 2 of and array , a vector is a 1D array

In [2]:
# the first step of using numpy is tell to tell python how to use it
# for that import is the keyword and import means what package we are getting and package name
import numpy as np
# so everytime we cant write numpy.something like numpy.exe etc so shortform is numpy as np where n stands for num and p stands for py


In [3]:
print(np.cos(np.pi))
print(np.log(np.exp(5.2)))
print(np.sqrt(1.21))
# in logarithemic values there are lots of things like natural logarithm which has base e and while writing no need to specify it and for rest it has like base 10 and base 2 so have to specify


-1.0
5.2
1.1


In [4]:
# we can create numpy arrays by creating lists
vec = np.array([1,2,3])
#vec stands for the vector
print(vec)
# we can create numpy by converting lists of lists
mat = np.array([[1,2,1],[4,5,9],[1,8,9]])
# mat stands for the matrix and we if keep np.array(lists of lists) it converts to matrix
print('')
print(mat)
print('')
print(mat.T)
#T stands for the transpose

[1 2 3]

[[1 2 1]
 [4 5 9]
 [1 8 9]]

[[1 4 1]
 [2 5 8]
 [1 9 9]]


## in a range starting point is inclusive and ending point is exclusive

In [5]:
# lots of ways to create the numpy arrays
vec2 = np.arange(0,15)
# arange stands for the array range 
# in python normal range only stores start point, end point and the step size 
# arange stores each and every values in data type
print(vec2)
print('')
vec3 = np.arange(3,21,6)
print(vec3)
print('')
a = range(0,15)
print(a)
print('')
b = range(2,16,5)
print(b)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]

[ 3  9 15]

range(0, 15)

range(2, 16, 5)


# in  a linspace that is called linear spacing both starting point and ennding point are inclusive
## linspace creates evenly spaced numbers between 2 values

In [6]:
vec4 = np.linspace(0,5,10)
#3 points need to have in linspace 0-starting,5-ending,10-how many poits do we need to have in an array
print(vec4)
print('')
# here it is like 5-0/10-1 (5/9) = 0.5556
# now it prints like 0+0.5556 , 0.5556+0.5556 etc so like 0, 0+step,0+2*step etc
print(vec4.reshape(5,2))
#reshape is for change array shape into 5 rows 3 columns
# here in reshape 5*2 becomes 10 so reshape is possible
# if i keep 5*# becomes 15 gives an error cause in an array only 10 entries are there
# if i keep 10*1 it will comes because if we want one coulmn
vec4_reshaped = vec4.reshape(5,2)
print(vec4_reshaped)
# also here vec4.reshape is a function that return a matrix and if we want to store set vec4.reshape equla to something

[0.         0.55555556 1.11111111 1.66666667 2.22222222 2.77777778
 3.33333333 3.88888889 4.44444444 5.        ]

[[0.         0.55555556]
 [1.11111111 1.66666667]
 [2.22222222 2.77777778]
 [3.33333333 3.88888889]
 [4.44444444 5.        ]]
[[0.         0.55555556]
 [1.11111111 1.66666667]
 [2.22222222 2.77777778]
 [3.33333333 3.88888889]
 [4.44444444 5.        ]]


## NumPy fills values row by row:
### | Row | Values             |
### | --- | ------------------ |
### | 1   | `0.0`, `0.5555`    |
### | 2   | `1.1111`, `1.6667` |
### | 3   | `2.2222`, `2.7778` |
### | 4   | `3.3333`, `3.8889` |
### | 5   | `4.4444`, `5.0`    |


In [7]:
mat2 = np.zeros([5,3])
print(mat2)
mat3 = np.ones((3,5))
print('')
print(mat3)
mat4 = np.eye(5)
print('')
print(mat4)
# np.eye Creates an identity matrix where diagonal is 1 and rest all are 0's

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]

[[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]]

[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]]


In [8]:
# we can +-*/ arrays together if they're the right size
vec5 = np.arange(1,6)
vec6 = np.arange(3,8)
print(vec5)
print(vec6)
print(vec5+vec6)
print(vec5*vec6)
print(1/vec5)
print(np.sqrt(vec6))

[1 2 3 4 5]
[3 4 5 6 7]
[ 4  6  8 10 12]
[ 3  8 15 24 35]
[1.         0.5        0.33333333 0.25       0.2       ]
[1.73205081 2.         2.23606798 2.44948974 2.64575131]


In [9]:
# we can do matrix multiplication
print(mat)
print('')
print(vec)
print()
product = np.matmul(mat,vec)
print(product)

[[1 2 1]
 [4 5 9]
 [1 8 9]]

[1 2 3]

[ 8 41 44]


# These lines use linear algebra functions from NumPy.


In [10]:
print(np.linalg.solve(mat,product)) 
#This solves a system of linear equations.

#It solves:

#Ax=b

#where:

#A = mat
#x = unknown values
#b = product

#So NumPy finds x.

#Ax=b
print('')
print(np.linalg.inv(mat))
# np.linalg.inv(mat) Finds the inverse matrix.

#The inverse is like division for matrices.

[1. 2. 3.]

[[ 0.5         0.18518519 -0.24074074]
 [ 0.5        -0.14814815  0.09259259]
 [-0.5         0.11111111  0.05555556]]


In [11]:
# we can find the unique values in an array
vec7 = np.array(['blue','red','orange','purple','purple','orange','Red',6])
print(vec7)
print(np.unique(vec7))

['blue' 'red' 'orange' 'purple' 'purple' 'orange' 'Red' '6']
['6' 'Red' 'blue' 'orange' 'purple' 'red']


In [12]:
# we can also use numpy to generate samples of a random variable
rand_mat = np.random.rand(5,5) # uniform random variable
print(rand_mat)
rand_mat2 = np.random.randn(10,5) # standard normal random variable
print('')
print(rand_mat2)

[[0.44177144 0.75992982 0.09259593 0.84323047 0.28828672]
 [0.53676298 0.47780047 0.08480979 0.2688181  0.20520187]
 [0.78276704 0.94251559 0.57319505 0.15122079 0.97755433]
 [0.49412656 0.69971599 0.32815546 0.32688968 0.44561469]
 [0.33029285 0.58788444 0.77983816 0.08690751 0.13243624]]

[[-0.43535771 -0.64252887  0.72550571 -0.42732914  0.81335456]
 [-0.50047427 -0.72662111 -0.74435546  1.66241124  0.41104686]
 [ 0.53561945 -0.05475368 -0.49213188  0.4282001   1.34596629]
 [ 0.30753645  0.19918284 -0.14666112 -1.35938669  0.94740643]
 [-0.25348836 -0.60862527  0.32381715 -0.45204851 -0.46365997]
 [-0.63034156  1.86526455  1.63988355 -0.25342564 -0.59599909]
 [-1.11273048  0.19240421 -1.15886938 -1.18617248 -0.5936323 ]
 [-0.71675743  0.81590635  0.78420936 -0.10561761  0.05353724]
 [-1.01331209 -1.43793938 -0.96902935  1.12348521 -0.24453879]
 [-0.81389836  0.20705644  1.32244819 -1.03272158 -0.06042499]]


In [13]:
# we can also use numpy for statistical tools on arrays
print(np.mean(rand_mat))
print(np.std(rand_mat2))

0.46553287844029045
0.8297858694935039


In [14]:
print(np.min(rand_mat))
print(np.max(rand_mat2))

0.08480979263680388
1.8652645471528218


# accesing entires to numpy arrays and matrices to acees individual entries

In [15]:
# how to access indvidual entry of vector just using the sqaure brackets
rand_vec = np.random.randn(19)
print(rand_vec)
print(rand_vec[6])


[ 1.78928754  1.18710605 -0.49990286 -0.55972629 -0.92923227 -0.68384106
 -1.19226979  0.19388584  1.80626405 -1.19321109  0.55794318  2.2902755
 -0.40732428  0.35847476 -0.87776448 -1.72814897  0.7089938   0.43996466
 -0.24994858]
-1.192269793186548



### rand_vec = np.random.randn(19)


### as:

### > “Give me 19 random numbers.”

Example:

array([ 0.3, -1.2, 0.7, 1.5, -0.4, ... ])


You get:

* some positive numbers
* some negative numbers
* mostly numbers close to `0`

---

## Breaking the line into pieces

### `np`

NumPy library.

---

### `random`

Random-number section of NumPy.

---

### `randn`

Means:

> random numbers from a “normal distribution”

You can simply think:

> “random numbers around 0”

---

### `(19)`

Means:

> generate 19 numbers

So:


np.random.randn(19)


means:

> “Generate 19 random numbers around 0.”

---

## Small example


np.random.randn(5)


might give:

array([ 0.2, -1.1, 0.5, 1.7, -0.3 ])


That is 5 random numbers.

---

## Why negative numbers appear

Because these numbers come from a bell-curve distribution centered at `0`.

Visual idea:


-3  -2  -1   0   1   2   3
             ^
          most numbers


Most values are near `0`,
fewer values are far away.


In [16]:
#access multiple entries at same time using numpy
print(rand_vec[4:9])

[-0.92923227 -0.68384106 -1.19226979  0.19388584  1.80626405]


In [17]:
# access multiple entries that are non consecutive using np array
print(np.arange(0,15,3))
print(rand_vec[np.arange(0,15,3)])
# here 0- starting point , 15- ending point and 3 is jump by 3 multiples and 15 is exculsive

[ 0  3  6  9 12]
[ 1.78928754 -0.55972629 -1.19226979 -1.19321109 -0.40732428]


In [18]:
print(rand_mat)
print(rand_mat[1][2])
print(rand_mat[1,2])

[[0.44177144 0.75992982 0.09259593 0.84323047 0.28828672]
 [0.53676298 0.47780047 0.08480979 0.2688181  0.20520187]
 [0.78276704 0.94251559 0.57319505 0.15122079 0.97755433]
 [0.49412656 0.69971599 0.32815546 0.32688968 0.44561469]
 [0.33029285 0.58788444 0.77983816 0.08690751 0.13243624]]
0.08480979263680388
0.08480979263680388


In [19]:
print(rand_mat[0:2,1:3])

[[0.75992982 0.09259593]
 [0.47780047 0.08480979]]


In [20]:
print(rand_vec)
rand_vec[3:5] = 4
print('')
print(rand_vec)
rand_vec[3:5]=[1,2]
print('')
print(rand_vec)
#first printing normal rand_vect 
#then here 3 is inclusive so index 3 = 4 and 5 is exclusive so index 4 = 4 
# printijng changes the index values
#then [3:5] equate to [1,2]
#then print again the index values of 3 and 4 are changes

[ 1.78928754  1.18710605 -0.49990286 -0.55972629 -0.92923227 -0.68384106
 -1.19226979  0.19388584  1.80626405 -1.19321109  0.55794318  2.2902755
 -0.40732428  0.35847476 -0.87776448 -1.72814897  0.7089938   0.43996466
 -0.24994858]

[ 1.78928754  1.18710605 -0.49990286  4.          4.         -0.68384106
 -1.19226979  0.19388584  1.80626405 -1.19321109  0.55794318  2.2902755
 -0.40732428  0.35847476 -0.87776448 -1.72814897  0.7089938   0.43996466
 -0.24994858]

[ 1.78928754  1.18710605 -0.49990286  1.          2.         -0.68384106
 -1.19226979  0.19388584  1.80626405 -1.19321109  0.55794318  2.2902755
 -0.40732428  0.35847476 -0.87776448 -1.72814897  0.7089938   0.43996466
 -0.24994858]


In [21]:
print(rand_mat)
rand_mat[1:3,3:5] = 0
print('')
print(rand_mat)

[[0.44177144 0.75992982 0.09259593 0.84323047 0.28828672]
 [0.53676298 0.47780047 0.08480979 0.2688181  0.20520187]
 [0.78276704 0.94251559 0.57319505 0.15122079 0.97755433]
 [0.49412656 0.69971599 0.32815546 0.32688968 0.44561469]
 [0.33029285 0.58788444 0.77983816 0.08690751 0.13243624]]

[[0.44177144 0.75992982 0.09259593 0.84323047 0.28828672]
 [0.53676298 0.47780047 0.08480979 0.         0.        ]
 [0.78276704 0.94251559 0.57319505 0.         0.        ]
 [0.49412656 0.69971599 0.32815546 0.32688968 0.44561469]
 [0.33029285 0.58788444 0.77983816 0.08690751 0.13243624]]


In [22]:
# we can also edit original matrix 
sub_mat = rand_mat[0:2,0:3] 
print(sub_mat)
sub_mat[:]=3
print(sub_mat)


[[0.44177144 0.75992982 0.09259593]
 [0.53676298 0.47780047 0.08480979]]
[[3. 3. 3.]
 [3. 3. 3.]]


In [23]:
print(rand_mat)

[[3.         3.         3.         0.84323047 0.28828672]
 [3.         3.         3.         0.         0.        ]
 [0.78276704 0.94251559 0.57319505 0.         0.        ]
 [0.49412656 0.69971599 0.32815546 0.32688968 0.44561469]
 [0.33029285 0.58788444 0.77983816 0.08690751 0.13243624]]


In [24]:
sub_mat2 = rand_mat[0:2,0:3].copy()
sub_mat2[:]=99
print(sub_mat2)
print(rand_mat)


[[99. 99. 99.]
 [99. 99. 99.]]
[[3.         3.         3.         0.84323047 0.28828672]
 [3.         3.         3.         0.         0.        ]
 [0.78276704 0.94251559 0.57319505 0.         0.        ]
 [0.49412656 0.69971599 0.32815546 0.32688968 0.44561469]
 [0.33029285 0.58788444 0.77983816 0.08690751 0.13243624]]


Suppose `rand_mat` is:

rand_mat = np.array([
    [10,20,30],
    [40,50,60],
    [70,80,90]
])

Now this line:

sub_mat2 = rand_mat[0:2,0:3].copy()


means:

* rows from `0` to `2` (2 not included)
* columns from `0` to `3` (3 not included)

So NumPy selects:

[
 [10,20,30],
 [40,50,60]
]

because:

* rows selected → 0 and 1
* columns selected → 0,1,2


## Important part: `.copy()`

Without `.copy()`, NumPy creates a **view** (linked to original matrix).

With `.copy()`, NumPy creates a **separate independent matrix**.

---

# This line

sub_mat2[:] = 99


means:

> “Replace every element with 99.”

Result:


[
 [99,99,99],
 [99,99,99]
]


# So:


print(sub_mat2)


prints:


[
 [99 99 99]
 [99 99 99]
]

# What about `rand_mat`?

Because you used `.copy()`:


print(rand_mat)


still prints the original matrix unchanged:

[
 [10 20 30]
 [40 50 60]
 [70 80 90]
]


# Very important concept

## With `.copy()`

Original matrix stays safe.

## Without `.copy()`

Example:

sub_mat2 = rand_mat[0:2,0:3]
sub_mat2[:] = 99


Then `rand_mat` would ALSO change because slicing creates a view by default in NumPy.


# access and modifying numpy array entries with logicals

In [25]:
rand_vec = np.random.randn(15)
print(rand_vec)
print(rand_vec>0)
print(rand_vec[rand_vec>0])

[ 1.10403112  0.08514395  0.79301914 -2.6387296  -0.60116434  0.50299467
  0.64435145  1.54972481 -0.97654435  0.2521202  -0.3229943  -0.24228012
  0.03131739  0.01550669  1.38614667]
[ True  True  True False False  True  True  True False  True False False
  True  True  True]
[1.10403112 0.08514395 0.79301914 0.50299467 0.64435145 1.54972481
 0.2521202  0.03131739 0.01550669 1.38614667]


In [26]:
print(rand_mat2)
print(rand_mat2[rand_mat2>0])
#here it is not deleting entire row and entire column its just deleting an individual

[[-0.43535771 -0.64252887  0.72550571 -0.42732914  0.81335456]
 [-0.50047427 -0.72662111 -0.74435546  1.66241124  0.41104686]
 [ 0.53561945 -0.05475368 -0.49213188  0.4282001   1.34596629]
 [ 0.30753645  0.19918284 -0.14666112 -1.35938669  0.94740643]
 [-0.25348836 -0.60862527  0.32381715 -0.45204851 -0.46365997]
 [-0.63034156  1.86526455  1.63988355 -0.25342564 -0.59599909]
 [-1.11273048  0.19240421 -1.15886938 -1.18617248 -0.5936323 ]
 [-0.71675743  0.81590635  0.78420936 -0.10561761  0.05353724]
 [-1.01331209 -1.43793938 -0.96902935  1.12348521 -0.24453879]
 [-0.81389836  0.20705644  1.32244819 -1.03272158 -0.06042499]]
[0.72550571 0.81335456 1.66241124 0.41104686 0.53561945 0.4282001
 1.34596629 0.30753645 0.19918284 0.94740643 0.32381715 1.86526455
 1.63988355 0.19240421 0.81590635 0.78420936 0.05353724 1.12348521
 0.20705644 1.32244819]


In [27]:
print(rand_vec)
print('')
rand_vec[rand_vec>0.5]=-5
print(rand_vec)

[ 1.10403112  0.08514395  0.79301914 -2.6387296  -0.60116434  0.50299467
  0.64435145  1.54972481 -0.97654435  0.2521202  -0.3229943  -0.24228012
  0.03131739  0.01550669  1.38614667]

[-5.          0.08514395 -5.         -2.6387296  -0.60116434 -5.
 -5.         -5.         -0.97654435  0.2521202  -0.3229943  -0.24228012
  0.03131739  0.01550669 -5.        ]


# save numpy arrays to hard drive

In [28]:
#for saving we can use numpy.save or np.save
#but np.save only saves single Numpy array
np.save('saved_file_name',rand_mat2)

In [29]:
#for saving more than one we have to use savez and z stands for zip
np.savez('zipped_file_name',rand_mat = rand_mat,rand_mat2=rand_mat2)
# the first rand_mat to the left of '=' sign is going to be a string inside of the zip file rand_mat that is right side of '=' sign is what numpy array that i actually want to sav



# loading the zip files

In [31]:
loaded_vec = np.load('saved_file_name.npy')
loaded_zip = np.load('zipped_file_name.npz')
print(loaded_vec)
# this give the rand_mat2
print('')
print(loaded_zip)
#this is giving some werid location

[[-0.43535771 -0.64252887  0.72550571 -0.42732914  0.81335456]
 [-0.50047427 -0.72662111 -0.74435546  1.66241124  0.41104686]
 [ 0.53561945 -0.05475368 -0.49213188  0.4282001   1.34596629]
 [ 0.30753645  0.19918284 -0.14666112 -1.35938669  0.94740643]
 [-0.25348836 -0.60862527  0.32381715 -0.45204851 -0.46365997]
 [-0.63034156  1.86526455  1.63988355 -0.25342564 -0.59599909]
 [-1.11273048  0.19240421 -1.15886938 -1.18617248 -0.5936323 ]
 [-0.71675743  0.81590635  0.78420936 -0.10561761  0.05353724]
 [-1.01331209 -1.43793938 -0.96902935  1.12348521 -0.24453879]
 [-0.81389836  0.20705644  1.32244819 -1.03272158 -0.06042499]]

NpzFile 'zipped_file_name.npz' with keys: rand_mat, rand_mat2



# NpzFile 'zipped_file_name.npz' with keys: rand_mat, rand_mat2 to access this


In [33]:
print(loaded_zip['rand_mat'])
print('')
print(loaded_zip['rand_mat2'])

[[3.         3.         3.         0.84323047 0.28828672]
 [3.         3.         3.         0.         0.        ]
 [0.78276704 0.94251559 0.57319505 0.         0.        ]
 [0.49412656 0.69971599 0.32815546 0.32688968 0.44561469]
 [0.33029285 0.58788444 0.77983816 0.08690751 0.13243624]]

[[-0.43535771 -0.64252887  0.72550571 -0.42732914  0.81335456]
 [-0.50047427 -0.72662111 -0.74435546  1.66241124  0.41104686]
 [ 0.53561945 -0.05475368 -0.49213188  0.4282001   1.34596629]
 [ 0.30753645  0.19918284 -0.14666112 -1.35938669  0.94740643]
 [-0.25348836 -0.60862527  0.32381715 -0.45204851 -0.46365997]
 [-0.63034156  1.86526455  1.63988355 -0.25342564 -0.59599909]
 [-1.11273048  0.19240421 -1.15886938 -1.18617248 -0.5936323 ]
 [-0.71675743  0.81590635  0.78420936 -0.10561761  0.05353724]
 [-1.01331209 -1.43793938 -0.96902935  1.12348521 -0.24453879]
 [-0.81389836  0.20705644  1.32244819 -1.03272158 -0.06042499]]


# instead of using python specific npy or npz we can also save simgle numpy array into text file
# if we use txt we can only stiore one array at a time

In [41]:
np.savetxt('text_file_name.txt',rand_mat,delimiter=",") #A delimiter is a character used to separate values.
rand_mat_txt =np.loadtxt('text_file_name.txt',delimiter=",")
print(rand_mat)
print('')
print(rand_mat_txt)

[[3.         3.         3.         0.84323047 0.28828672]
 [3.         3.         3.         0.         0.        ]
 [0.78276704 0.94251559 0.57319505 0.         0.        ]
 [0.49412656 0.69971599 0.32815546 0.32688968 0.44561469]
 [0.33029285 0.58788444 0.77983816 0.08690751 0.13243624]]

[[3.         3.         3.         0.84323047 0.28828672]
 [3.         3.         3.         0.         0.        ]
 [0.78276704 0.94251559 0.57319505 0.         0.        ]
 [0.49412656 0.69971599 0.32815546 0.32688968 0.44561469]
 [0.33029285 0.58788444 0.77983816 0.08690751 0.13243624]]
